src/org/zaproxy/zap/extension/proxies/ExtensionProxies.java

```diff
@@ -0,0 +1,233 @@
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); 
+ * you may not use this file except in compliance with the License. 
+ * You may obtain a copy of the License at 
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0 
+ *   
+ * Unless required by applicable law or agreed to in writing, software 
+ * distributed under the License is distributed on an "AS IS" BASIS, 
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
+ * See the License for the specific language governing permissions and 
+ * limitations under the License. 
+ */
+package org.zaproxy.zap.extension.proxies;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.MalformedURLException;
+import java.net.ServerSocket;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.control.Control;
+import org.parosproxy.paros.core.proxy.ProxyServer;
+import org.parosproxy.paros.extension.ExtensionAdaptor;
+import org.parosproxy.paros.extension.ExtensionHook;
+import org.parosproxy.paros.extension.OptionsChangedListener;
+import org.parosproxy.paros.extension.history.ExtensionHistory;
+import org.parosproxy.paros.model.OptionsParam;
+
+public class ExtensionProxies extends ExtensionAdaptor implements OptionsChangedListener {
+
+    public static final String NAME = "ExtensionProxies";
+    public static final String ZAP_PROXY_THREAD_PREFIX = "ZAP-";
+
+    private ProxiesParam proxiesParam = null;
+    private OptionsProxiesPanel optionsProxiesPanel = null;
+    private Map<String, ProxyServer> proxyServers = new HashMap<String, ProxyServer>();
+
+    private static Logger log = Logger.getLogger(ExtensionProxies.class);
+
+    public ExtensionProxies() {
+        super();
+        initialize();
+    }
+
+    private void initialize() {
+        this.setName(NAME);
+        this.setOrder(310);
+    }
+
+    @Override
+    public void init() {
+    }
+
+    @Override
+    public String getUIName() {
+        return Constant.messages.getString("proxies.name");
+    }
+
+    @Override
+    public void hook(ExtensionHook extensionHook) {
+        super.hook(extensionHook);
+
+        extensionHook.addOptionsParamSet(this.getParam());
+        extensionHook.addOptionsChangedListener(this);
+
+        if (getView() != null) {
+            extensionHook.getHookView().addOptionPanel(getOptionsProxiesPanel());
+        }
+
+        extensionHook.addApiImplementor(new ProxiesAPI(this));
+    }
+
+    @Override
+    public void unload() {
+        super.unload();
+        // Stop all of the running servers
+        for (Entry<String, ProxyServer> entry : proxyServers.entrySet()) {
+            stopProxyServer(entry.getValue());
+        }
+        proxyServers.clear();
+    }
+
+    @Override
+    public void optionsLoaded() {
```

Comment:
This could be done later (`start`?). Also, it might happen that if you set an alternative proxy with the same port as the main one the main one will not be started (which in daemon means ZAP will fail to start).